core/shared/src/main/scala/fs2/internal/CompileScope.scala

```diff
@@ -113,37 +113,108 @@ private[fs2] final class CompileScope[F[_], O] private (
       interruptible: Option[
         (Effect[F], ExecutionContext, FreeC[Algebra[F, O, ?], Unit])
       ]
-  ): F[CompileScope[F, O]] =
-    F.flatMap(state.modify2 { s =>
-      if (!s.open) (s, None)
-      else {
-        val newScopeId = new Token
-        val iCtx = interruptible.map {
-          case (effect, ec, whenInterrupted) =>
-            InterruptContext[F, O](
-              effect = effect,
-              ec = ec,
-              promise = Promise.unsafeCreate[F, Option[Throwable]](effect, ec),
-              ref = Ref.unsafeCreate[F, Option[Option[Throwable]]](None),
-              interruptRoot = newScopeId,
-              whenInterrupted = whenInterrupted
-            )
-        }
-        val scope = new CompileScope[F, O](newScopeId, Some(self), iCtx.orElse(self.interruptible))
+  ): F[CompileScope[F, O]] = {
+
+    /**
+      * Creates a context for a new scope.
+      *
+      * We need to differentiate between three states:
+      *  The new scope is not interruptible -
+      *     It should respect the interrupt of the current scope. But it should not
+      *     close the listening on parent scope close when the new scope will close.
+      *
+      *  The new scope is interruptible but this scope is not interruptible -
+      *     This is a new interrupt root that can be only interrupted from bottom. Create new interrupt context.
+      *
+      *  The new scope is interruptible as well as this scope is interruptible -
+      *     This is a new interrupt root that can be interrupted from bottom or as a result of
+      *     interrupting this scope. But it should not propagate its own interruption to this scope.
+      *
+      */
+    def createScopeContext: F[(Option[InterruptContext[F, O]], Token)] =
+      interruptible match {
```

Comment:
I see lot of noise/redundancies in this code. Cannot we simplify it ? i.e. effect, ec, promise, ref, seems to be same or fold over option, and real different is when parent is interruptible and new scope shall be root of another interruptible context.